### PR TITLE
Allow turning shared library building off

### LINKED ISSUE #740: Problems when cross compiling for ARM
Hello,

I've got several problem when trying to cross compile Tango for ARM platform:
gcc version 4.7.3 (Sourcery CodeBench Lite 2013.05-40) 

My cmake build:
```
[debian9-devel.pons] > cmake .. -DCMAKE_CXX_COMPILER:FILEPATH=/segfs/linux/xilinx_sdk/arm-toolchain/bin/arm-xilinx-linux-gnueabi-gcc -DIDL_BASE=/users/pons/tango/tango-idl/install -DOMNI_BASE=/segfs/tango/ORB/omniORB4.2.1_p1/arm_xilinx -DZMQ_BASE=/segfs/tango/transport/zmq4.0.5/arm_xilinx -DIDL_BASE=/users/pons/tango/tango-idl -DTANGO_JPEG_MMX=OFF -DBUILD_TESTING=OFF -DCMAKE_INSTALL_PREFIX=/segfs/tango/release/arm_xilinx/9.3.4-rc6
System TANGO_HOST=acudebian7:10000
No build type specified - default is Release
Using IDL_BASEIDL_BASE=/users/pons/tango/tango-idl
Using OMNI_BASE=/segfs/tango/ORB/omniORB4.2.1_p1/arm_xilinx
Using ZMQ_BASE=/segfs/tango/transport/zmq4.0.5/arm_xilinx
-- Looking for C++ include zmq.hpp
-- Looking for C++ include zmq.hpp - found
-- Check if zmq::socket has a disconnect method: FALSE
Verifying /segfs/tango/ORB/omniORB4.2.1_p1/arm_xilinx/bin/omniidl
omniidl version 1.0
found GNU compiler ...
setup for release build ...
Remove tango.h, tangoSK.cpp and tangoDynSK.cpp from idl if exist
Generate tango.h, tangoSK.cpp and tangoDynSK.cpp from idl
Using OMNIIDL_PATH=/segfs/tango/ORB/omniORB4.2.1_p1/arm_xilinx/bin/
Using IDL=/users/pons/tango/tango-idl/include
-- OMNIIDL_PATH: /segfs/tango/ORB/omniORB4.2.1_p1/arm_xilinx/bin/
-- CMAKE_CURRENT_BINARY_DIR: /users/pons/tango/cppTango/build/cppapi/server/idl
-- CMAKE_CURRENT_SOURCE_DIR: /users/pons/tango/cppTango/cppapi/server/idl
-- CMAKE_INCLUDE_PATH: 
Applying enhancement /users/pons/tango/cppTango/cppapi/server/idl/EnhanceAttributeValue_4
Applying enhancement /users/pons/tango/cppTango/cppapi/server/idl/EnhanceAttributeValue_5
Applying enhancement /users/pons/tango/cppTango/cppapi/server/idl/EnhanceDevPipeData
-- Configuring done
-- Generating done
-- Build files have been written to: /users/pons/tango/cppTango/build
```

**1st problem**: [log4tango] It seems that cmake does not detect stringstream and that the file Portability.hh does not work when stringstream is not found.

```
                 from /users/pons/tango/cppTango/log4tango/src/Appender.cpp:29:
/users/pons/tango/cppTango/log4tango/include/log4tango/Portability.hh:56:26: error: using typedef-name 'std::ostringstream' after 'class'
```

However, the cross compiler used supports well stringstream and by modifying the config.h by hand, it solved the problem.
log4tango/include/log4tango/config.h 

```
/* define if the compiler has stringstream */
/* #undef LOG4TANGO_HAVE_SSTREAM */
```
changed to:
```
/* define if the compiler has stringstream */
#define LOG4TANGO_HAVE_SSTREAM
```

**2nd problem**: [tango] How to disable the building of shared library ?
I don't have omniORB and zmq compiled in shared, so the compilation of tango SHARED failed and prevent the build process to end properly. To make the build process working, I had to comment in configure/cmake_linux.cmake:

```
#add_library(tango SHARED    $<TARGET_OBJECTS:log4tango_objects>
#                            $<TARGET_OBJECTS:client_objects>
#                            $<TARGET_OBJECTS:idl_objects>
#                            $<TARGET_OBJECTS:jpeg_objects>
#                            $<TARGET_OBJECTS:jpeg_mmx_objects>
#                            $<TARGET_OBJECTS:server_objects>)
#target_link_libraries(tango PUBLIC ${ZMQ_PKG_LIBRARIES} ${OMNIORB_PKG_LIBRARIES} ${OMNICOS_PKG_LIBRARIES} ${OMNIDYN_PKG_LIBRARIES} ${CMAKE_DL_LIBS})
#target_compile_options(tango PRIVATE -fPIC)
#target_include_directories(tango PUBLIC ${ZMQ_PKG_INCLUDE_DIRS} ${OMNIORB_PKG_INCLUDE_DIRS} ${OMNIDYN_PKG_INCLUDE_DIRS})

#target_compile_options(tango PUBLIC ${ZMQ_PKG_CFLAGS_OTHER} ${OMNIORB_PKG_CFLAGS_OTHER} ${OMNICOS_PKG_CFLAGS_OTHER} ${OMNIDYN_PKG_CFLAGS_OTHER})
#target_compile_definitions(tango PUBLIC _REENTRANT)

#set_target_properties(tango PROPERTIES
#                      VERSION ${LIBRARY_VERSION}
#                      SOVERSION ${SO_VERSION})
```

**3rd problem** : [log4tango] test was not disabled in log4tango even when using BUILD_TESTING=OFF
I had to comment the following line in log4tango/CMakeList.txt

```
#if (NOT WIN32)
#   add_subdirectory(tests)
#endif(NOT WIN32)
```

### PR DESCRIPTION
**Pull request by t-b**
Close #740.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -52,14 +53,19 @@ add_subdirectory("log4tango")
 add_subdirectory("cppapi")
 
 if(BUILD_TESTING)
+  if(BUILD_SHARED_LIBS)
     add_subdirectory("cpp_test_suite")
+  else()
+    message(WARNING "Not building the tests, because that is only possible when BUILD_SHARED_LIBS is ON")
+    SET(BUILD_TESTING OFF)
```
Comment (mliszcz) [inline CMakeLists.txt]: What is the reason for skipping tests in this case? Why can't we just link the tests statically with libtango.a?

**Review by mliszcz — COMMENTED** [review]
Thanks for this change! I added one comment regarding the tests, could you please have a look?

File: CMakeLists.txt (hunk 1 above)
Comment (t-b) [reply CMakeLists.txt]: @mliszcz It does not work out of the box, that is the only reason. Should I get it working?

Comment (mliszcz) [reply CMakeLists.txt]: @t-b thanks for clarification. I was just wondering why it is not possible. No need to get it working IMO. I'd just change the warning message from "it is only possible when" to "it is currently supported only when", but I do not have a strong opinion here, it's up to you.

**Review by mliszcz — DISMISSED** [review]

**Comment by mliszcz** [comment]
Maybe we should have a CI job for static build on Linux?

File: CMakeLists.txt (hunk 1 above)
Comment (t-b) [reply CMakeLists.txt]: Good point. I'll reword when rerolling.

**Comment by t-b** [comment]
@mliszcz I've added a travis job.

@mliszcz @bourtemb Any opinions about backporting? I would think that having that in 9.3-backports would be good. IMHO only the last two commits are needed.

**Review by mliszcz — APPROVED** [review]
LGTM.

As for the backport, I do not have any opinion here. Maybe just ask the users if they need this.

**Comment by JeanLucPons** [comment]
Hello,
Thanks for this. I tried but am I doing something wrong ?
Thanks for help.

```
cmake .. -DCMAKE_CXX_COMPILER:FILEPATH=/segfs/linux/xilinx_sdk/arm-toolchain/bin/arm-xilinx-linux-gnueabi-g++ -DIDL_BASE=/users/pons/tango/tango-idl/install -DOMNI_BASE=/segfs/tango/ORB/omniORB4.2.1_p1/arm_xilinx -DZMQ_BASE=/segfs/tango/transport/zmq4.0.5/arm_xilinx -DIDL_BASE=/users/pons/tango/tango-idl -DTANGO_JPEG_MMX=OFF -DBUILD_SHARED_LIBS=OFF -DBUILD_TESTING=OFF -DCMAKE_INSTALL_PREFIX=/segfs/tango/release/arm_xilinx/PR758
```

```
- CMake: version 3.7.2
-- Target platform: Linux 32-bit
-- C++ Compiler: GNU with version 4.7.3
-- C Compiler: GNU with version 6.3.0
-- Build type: Release
-- Environment: TANGO_HOST=acudebian7:10000
Using IDL_BASEIDL_BASE=/users/pons/tango/tango-idl
Using OMNI_BASE=/segfs/tango/ORB/omniORB4.2.1_p1/arm_xilinx
Using ZMQ_BASE=/segfs/tango/transport/zmq4.0.5/arm_xilinx
-- Looking for C++ include zmq.hpp
-- Looking for C++ include zmq.hpp - found
-- Check if zmq::socket has a disconnect method: FALSE
Verifying /segfs/tango/ORB/omniORB4.2.1_p1/arm_xilinx/bin/omniidl
omniidl version 1.0
found GNU compiler ...
setup for release build ...
-- Check if the compiler supports C++14 relaxed constexpr: FALSE
Generate tango.h, tangoSK.cpp and tangoDynSK.cpp from idl
Using OMNIIDL_PATH=/segfs/tango/ORB/omniORB4.2.1_p1/arm_xilinx/bin/
Using IDL=/users/pons/tango/tango-idl/include
-- Configuring done
CMake Error in CMakeLists.txt:
  Target "tango-static" requires the language dialect "CXX14" , but CMake
  does not know the compile flags to use to enable it.
```

**Comment by t-b** [comment]
@JeanLucPons CI passes, and it has a new job with static configuration, see [1].

- You are using different compilers for C and C++, this does look odd.
- I think cmake tries to tell you that your compiler is too old. You can try passing `-DCMAKE_CXX_STANDARD=11` which might still work, but we will be moving towards C++14. So using something newer than g++ 4.7 is definitly more future proof.

[1]: https://travis-ci.org/github/tango-controls/cppTango/jobs/723093576

**Comment by JeanLucPons** [comment]
Unfortunately I cannot update the compiler. The toolchain has been built by a manufacturer for a specific hardware and cannot be updated. I tried to use -DCMAKE_CXX_STANDARD=11 and the cmake ended successfully. However the make wants to build MMX although it was disabled (-DTANGO_JPEG_MMX=OFF).

```
[debian9-devel.pons] > cmake .. -DCMAKE_CXX_COMPILER:FILEPATH=/segfs/linux/xilinx_sdk/arm-toolchain/bin/arm-xilinx-linux-gnueabi-g++ -DIDL_BASE=/users/pons/tango/tango-idl/install -DOMNI_BASE=/segfs/tango/ORB/omniORB4.2.1_p1/arm_xilinx -DZMQ_BASE=/segfs/tango/transport/zmq4.0.5/arm_xilinx -DIDL_BASE=/users/pons/tango/tango-idl -DTANGO_JPEG_MMX=OFF -DBUILD_SHARED_LIBS=OFF -DBUILD_TESTING=OFF -DCMAKE_CXX_STANDARD=11 -DCMAKE_INSTALL_PREFIX=/segfs/tango/release/arm_xilinx/PR758-- CMake: version 3.7.2
-- Target platform: Linux 32-bit
-- C++ Compiler: GNU with version 4.7.3
-- C Compiler: GNU with version 6.3.0
-- Build type: Release
-- Environment: TANGO_HOST=acudebian7:10000
Using IDL_BASEIDL_BASE=/users/pons/tango/tango-idl
Using OMNI_BASE=/segfs/tango/ORB/omniORB4.2.1_p1/arm_xilinx
Using ZMQ_BASE=/segfs/tango/transport/zmq4.0.5/arm_xilinx
-- Looking for C++ include zmq.hpp
-- Looking for C++ include zmq.hpp - found
-- Check if zmq::socket has a disconnect method: FALSE
Verifying /segfs/tango/ORB/omniORB4.2.1_p1/arm_xilinx/bin/omniidl
omniidl version 1.0
found GNU compiler ...
setup for release build ...
-- Check if the compiler supports C++14 relaxed constexpr: FALSE
Generate tango.h, tangoSK.cpp and tangoDynSK.cpp from idl
Using OMNIIDL_PATH=/segfs/tango/ORB/omniORB4.2.1_p1/arm_xilinx/bin/
Using IDL=/users/pons/tango/tango-idl/include
-- Configuring done
-- Generating done
-- Build files have been written to: /users/pons/tango/cppTango.PR758/build
[debian9-devel.pons] > 
[debian9-devel.pons] > make
[  0%] Building CXX object cppapi/server/jpeg_mmx/CMakeFiles/jpeg_mmx_objects.dir/jpeg_color_mmx.cpp.o
arm-xilinx-linux-gnueabi-g++: error: unrecognized command line option '-fstack-protector-strong'
cppapi/server/jpeg_mmx/CMakeFiles/jpeg_mmx_objects.dir/build.make:62: recipe for target 'cppapi/server/jpeg_mmx/CMakeFiles/jpeg_mmx_objects.dir/jpeg_color_mmx.cpp.o' failed
```

**Comment by mliszcz** [comment]
Hi @JeanLucPons, let me clarify a few points here:

* cppTango (tango-9-lts branch and also this PR) requires C++14. `CMAKE_CXX_STANDARD` is here to allow compiling with *newer* C++ standard versions. Things generally won't work if you pass C++11 there. With your compiler you should use one of 9.3.x releases.

* The flag TANGO_JPEG_MMX if for disabling use of MMX extensions in our internal jpeg_mmx library. This basically disables `-mmmx` switch. The library is still compiled, as you noticed.

* The compilation is failing due to your compiler not supporting `-fstack-protector-strong` flag. In GCC this flag is only available since 4.9.0 release. We do not support older GCC releases as they do not support most of  [C++14 features](https://gcc.gnu.org/projects/cxx-status.html#cxx14) (see first bullet). As of today, the oldest supported and tested in CI version of GCC is 4.9.2. To work around this particular problem, you may try building in Debug mode which should not include any hardening flags (`-DCMAKE_BUILD_TYPE=Debug`).

**Comment by JeanLucPons** [comment]
OK thanks.
There is no interest for me to compile in debug mode as I search maximum performance for running server on small embedded devices (Spark BPM). For the moment the 9.3.4-rc6 is working well and I hope it will work for the next years... We have at the ESRF lots of BPM (and BLM) spark devices that require embedded Tango.

**Comment by t-b** [comment]
@JeanLucPons Will make a backport PR available.

**Comment by t-b** [comment]
@JeanLucPons See https://github.com/tango-controls/cppTango/pull/763 for the backport PR.

**Review by bourtemb — APPROVED** [review]
Looks good to me.
Thanks!

**Pull request merged**